spark/src/main/scala/io/delta/sharing/spark/DeltaSharingSource.scala

```diff
@@ -0,0 +1,516 @@
+   * version is valid by comparing with previous version.
+   * @param indexedFile The last indexed file used to build offset from.
+   * @param version Previous offset table version.
+   * @param isStartingVersion Whether previous offset is starting version or not.
+   */
+  private def buildOffsetFromIndexedFile(
+    indexedFile: IndexedFile,
+    version: Long,
+    isStartingVersion: Boolean): Option[DeltaSharingSourceOffset] = {
+    val IndexedFile(v, i, _, isLastFileInVersion) = indexedFile
+    assert(v >= version,
+      s"buildOffsetFromIndexedFile receives an invalid version: $v (expected: >= $version), " +
+        s"tableId: $tableId")
+
+    // If the last file in previous batch is the last file of that version, automatically bump
+    // to next version to skip accessing that version file altogether.
+    if (isLastFileInVersion) {
+      // isStartingVersion must be false here as we have bumped the version.
+      Some(DeltaSharingSourceOffset(DeltaSharingSourceOffset.VERSION_1, tableId, v + 1, index = -1,
+        isStartingVersion = false))
+    } else {
+      // isStartingVersion will be true only if previous isStartingVersion is true and the next file
+      // is still at the same version (i.e v == version).
+      Some(DeltaSharingSourceOffset(DeltaSharingSourceOffset.VERSION_1, tableId, v, i,
+        isStartingVersion = v == version && isStartingVersion))
+    }
+  }
+}
+
+/**
+ * A streaming source for a Delta Sharing table.
+ *
+ * When a new stream is started, delta sharing starts by constructing a [[RemoteDeltaSnapshot]]
+ * at the current version of the table. This snapshot is broken up into batches until
+ * all existing data has been processed. Subsequent processing is done by tailing
+ * the change log looking for new data. This results in the streaming query returning
+ * the same answer as a batch query that had processed the entire dataset at any given point.
+ */
+case class DeltaSharingSource(
+  spark: SparkSession,
+  deltaLog: RemoteDeltaLog,
+  options: DeltaSharingOptions)
+  extends DeltaSharingSourceBase {
+
+  /** A check on the source table that disallows deletes on the source data. */
+  private val ignoreChanges = options.ignoreChanges
+
+  /** A check on the source table that disallows commits that only include deletes to the data. */
+  private val ignoreDeletes = options.ignoreDeletes || ignoreChanges
+
+  // This is checked before creating DeltaSharingSource
+  assert(schema.nonEmpty)
+
+  protected val tableId = snapshot.metadata.id
+
+  private var previousOffset: DeltaSharingSourceOffset = null
+
+  protected def verifyStreamHygieneAndFilterAddFiles(
+    tableFiles: DeltaTableFiles): Seq[AddFileForCDF] = {
+    if (!tableFiles.removeFiles.isEmpty) {
+      val groupedRemoveFiles = tableFiles.removeFiles.groupBy(r => r.version)
+      val groupedAddFiles = tableFiles.addFiles.groupBy(a => a.version)
```

Comment:
Question: It seems groupBy will do a lot more work than is needed... right? 
technically we don't really need groups for verification, we just need info such as whether a version has add and/or remove files... right?

Reply:
updated to Set.